torch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,90 @@
+import math
+from numbers import Number
+
+import torch
+from torch.autograd import Variable
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all
+
+class MultivariateNormal(Distribution):
+    r"""
+    Creates a multivariate normal (also called Gaussian) distribution
+    parameterized by a mean vector and a covariance matrix.
+    
+    The multivariate normal distribution can be parameterized either 
+    in terms of a positive definite covariance matrix :math:`\mathbf{\Sigma}` 
+    or a lower-triangular matrix :math:`\mathbf{L}` such that 
+    :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top` as obtained via e.g. 
+    Cholesky decomposition of the covariance.
+
+    Example:
+
+        >>> m = MultivariateNormal(torch.zeros(2), torch.eye(2))
+        >>> m.sample()  # normally distributed with mean=`[0,0]` and covariance_matrix=`I`
+        -0.2102
+        -0.5429
+        [torch.FloatTensor of size 2]
+
+    Args:
+        loc (Tensor or Variable): mean of the distribution
+        covariance_matrix (Tensor or Variable): covariance matrix (sigma positive-definite).
+        scale_tril (Tensor or Variable): lower-triangular factor of covariance.
+        
+    Note:
+        Only one of `covariance_matrix` or `scale_tril` can be specified.
+        
+    """
+    params = {'loc': constraints.real,
+              'covariance_matrix': constraints.positive_definite,
+              'scale_tril': constraints.lower_triangular }
+    support = constraints.real
+    has_rsample = True
+
+    def __init__(self, loc, covariance_matrix=None, scale_tril=None):
+        batch_shape, event_shape = loc.shape[:-1], loc.shape[-1:]
+        if covariance_matrix is not None and scale_tril is not None:
+            raise ValueError("Either covariance matrix or scale_tril may be specified, not both.")
+        if covariance_matrix is None and scale_tril is None:
+            raise ValueError("One of either covariance matrix or scale_tril must be specified")
+        if scale_tril is None:
```

Comment:
Neeraj made this cool decorator called `@lazy_property` that could allow you to create `scale_tril` only if it does not exist, which would avoid unnecessary work in some cases. You could use it as follows:
```py
class MultivariateNormal(Distribution):
    def __init__(...):
        ...
        if scale_tril is not None:
            self.scale_tril = scale_tril
            # leave .covariance_matrix unset
        else:
            self.covariance_matrix = covariance_matrix
            # leave .scale_tril unset
    ...
    @lazy_property
    def scale_tril(self):
        return torch.potrf(self.covariance_matrix, upper=False)
    @lazy_property
    def covariance_matrix(self):
        return torch.mm(scale_tril, scale_tril.t())
```